File: Advanced Robotics Course/Chapter2/Lesson3/Chapter2_Lesson3.java
import java.util.*;
import static java.lang.Math.*;

public class LatticePlanner {

    static final int NX = 50, NY = 30, N_THETA = 16;
    static final double DX = 1.0, DY = 1.0, L = 2.0, V = 1.0, DT = 1.0;

    static int[][] occGrid = new int[NX][NY];

    static class State {
        int ix, iy, kth;
        State(int ix, int iy, int kth) {
            this.ix = ix; this.iy = iy; this.kth = kth;
        }
        @Override public boolean equals(Object o) {
            if (!(o instanceof State)) return false;
            State s = (State)o;
            return ix == s.ix && iy == s.iy && kth == s.kth;
        }
        @Override public int hashCode() {
            return ix * 73856093 ^ iy * 19349663 ^ kth * 83492791;
        }
    }

    static class Node implements Comparable<Node> {
        State s;
        double g, f;
        Node(State s, double g, double f) {
            this.s = s; this.g = g; this.f = f;
        }
        public int compareTo(Node other) {
            return Double.compare(this.f, other.f);
        }
    }

    static class Primitive {
        double dxRef, dyRef, dthRef, length;
        Primitive(double dx, double dy, double dth, double len) {
            dxRef = dx; dyRef = dy; dthRef = dth; length = len;
        }
    }

    static boolean inBounds(int ix, int iy) {
        return ix >= 0 && ix < NX && iy >= 0 && iy < NY;
    }

    static boolean isFree(int ix, int iy) {
        return inBounds(ix, iy) && occGrid[ix][iy] == 0;
    }

    static int angleIndex(double theta) {
        double twoPi = 2.0 * PI;
        theta = (theta + twoPi) % twoPi;
        int k = (int) Math.round(theta / twoPi * N_THETA) % N_THETA;
        return k;
    }

    static double angleFromIndex(int k) {
        return 2.0 * PI * k / N_THETA;
    }

    static List<Primitive> makePrimitives() {
        double[] steer = {-0.4, 0.0, 0.4};
        List<Primitive> prims = new ArrayList<>();
        for (double phi : steer) {
            double kappa = Math.tan(phi) / L;
            if (Math.abs(kappa) < 1e-6) {
                prims.add(new Primitive(V * DT, 0.0, 0.0, V * DT));
            } else {
                double dth = V * kappa * DT;
                double thEnd = dth;
                double dx = (1.0 / kappa) * Math.sin(thEnd);
                double dy = -(1.0 / kappa) * (Math.cos(thEnd) - 1.0);
                double length = Math.abs(dth / kappa);
                prims.add(new Primitive(dx, dy, dth, length));
            }
        }
        return prims;
    }

    static double heuristic(State s, State goal) {
        double x = s.ix * DX;
        double y = s.iy * DY;
        double xg = goal.ix * DX;
        double yg = goal.iy * DY;
        return Math.hypot(x - xg, y - yg);
    }

    public static List<State> astar(State start, State goal) {
        List<Primitive> prims = makePrimitives();
        PriorityQueue<Node> open = new PriorityQueue<>();
        Map<State, Double> g = new HashMap<>();
        Map<State, State> parent = new HashMap<>();
        Set<State> closed = new HashSet<>();

        g.put(start, 0.0);
        parent.put(start, null);
        open.add(new Node(start, 0.0, heuristic(start, goal)));

        while (!open.isEmpty()) {
            Node cur = open.poll();
            if (closed.contains(cur.s)) continue;
            closed.add(cur.s);

            if (cur.s.ix == goal.ix && cur.s.iy == goal.iy) {
                List<State> path = new ArrayList<>();
                State s = cur.s;
                while (s != null) {
                    path.add(s);
                    s = parent.get(s);
                }
                Collections.reverse(path);
                return path;
            }

            double theta = angleFromIndex(cur.s.kth);
            for (Primitive p : prims) {
                double gx = cur.s.ix * DX;
                double gy = cur.s.iy * DY;
                double dx = Math.cos(theta) * p.dxRef - Math.sin(theta) * p.dyRef;
                double dy = Math.sin(theta) * p.dxRef + Math.cos(theta) * p.dyRef;
                double gx2 = gx + dx;
                double gy2 = gy + dy;
                double th2 = theta + p.dthRef;

                int ix2 = (int) Math.round(gx2 / DX);
                int iy2 = (int) Math.round(gy2 / DY);
                int kth2 = angleIndex(th2);
                if (!isFree(ix2, iy2)) continue;

                State s2 = new State(ix2, iy2, kth2);
                double newG = g.get(cur.s) + p.length;
                if (!g.containsKey(s2) || newG < g.get(s2)) {
                    g.put(s2, newG);
                    parent.put(s2, cur.s);
                    double f = newG + heuristic(s2, goal);
                    open.add(new Node(s2, newG, f));
                }
            }
        }
        return null;
    }

    public static void main(String[] args) {
        State start = new State(2, 2, angleIndex(0.0));
        State goal = new State(40, 20, angleIndex(0.0));
        List<State> path = astar(start, goal);
        if (path != null) {
            System.out.println("Found path of length " + path.size());
        } else {
            System.out.println("No path found.");
        }
    }
}
      

File: Advanced Robotics Course/Chapter2/Lesson3/Chapter2_Lesson3.py
import heapq
import math
import numpy as np

# Discretization parameters
DX = 1.0
DY = 1.0
N_THETA = 16  # number of discrete headings
DT = 1.0      # primitive duration
V = 1.0       # forward speed
L = 2.0       # wheelbase

# Occupancy grid: 0 = free, 1 = obstacle
# Here we just create a small empty grid as an example.
NX, NY = 50, 30
occ_grid = np.zeros((NX, NY), dtype=np.uint8)

def in_bounds(ix, iy):
    return 0 <= ix < NX and 0 <= iy < NY

def is_free(ix, iy):
    return in_bounds(ix, iy) and occ_grid[ix, iy] == 0

def angle_index(theta):
    """Map continuous angle to discrete index."""
    theta = (theta + 2.0 * math.pi) % (2.0 * math.pi)
    k = int(round(theta / (2.0 * math.pi) * N_THETA)) % N_THETA
    return k

def angle_from_index(k):
    return 2.0 * math.pi * k / N_THETA

# Precompute reference motion primitives at pose (0,0,0)
STEER_ANGLES = [-0.4, 0.0, 0.4]  # radians, simple left/straight/right
ref_primitives = []  # each entry: (dx, dy, dtheta, length)

for phi in STEER_ANGLES:
    kappa = math.tan(phi) / L  # curvature
    if abs(kappa) < 1e-6:
        # straight-line primitive
        dx = V * DT
        dy = 0.0
        dtheta = 0.0
        length = V * DT
    else:
        # circular-arc primitive
        dtheta = V * kappa * DT
        th_end = dtheta  # from 0 to dtheta
        dx = (1.0 / kappa) * math.sin(th_end)
        dy = -(1.0 / kappa) * (math.cos(th_end) - 1.0)
        length = abs(dtheta / kappa)
    ref_primitives.append((dx, dy, dtheta, length))

def successors(state):
    """
    state = (ix, iy, kth)
    Returns list of (next_state, cost).
    """
    ix, iy, kth = state
    theta = angle_from_index(kth)
    succs = []
    for dx_ref, dy_ref, dth_ref, length in ref_primitives:
        # transform reference primitive to global frame
        gx = ix * DX
        gy = iy * DY
        dx = math.cos(theta) * dx_ref - math.sin(theta) * dy_ref
        dy = math.sin(theta) * dx_ref + math.cos(theta) * dy_ref
        gx2 = gx + dx
        gy2 = gy + dy
        kth2 = angle_index(theta + dth_ref)
        ix2 = int(round(gx2 / DX))
        iy2 = int(round(gy2 / DY))

        if not is_free(ix2, iy2):
            continue

        cost = length  # use arc length as cost
        succs.append(((ix2, iy2, kth2), cost))
    return succs

def heuristic(state, goal_state):
    ix, iy, _ = state
    gx, gy, _ = goal_state
    x = ix * DX
    y = iy * DY
    xg = gx * DX
    yg = gy * DY
    return math.hypot(x - xg, y - yg)

def astar(start_state, goal_state):
    """
    A* search on the lattice.
    state = (ix, iy, kth)
    """
    open_heap = []
    g = {start_state: 0.0}
    parent = {start_state: None}
    f_start = heuristic(start_state, goal_state)
    heapq.heappush(open_heap, (f_start, start_state))

    closed = set()

    while open_heap:
        f_curr, curr = heapq.heappop(open_heap)
        if curr in closed:
            continue
        closed.add(curr)

        if (curr[0], curr[1]) == (goal_state[0], goal_state[1]):
            # reached goal cell (ignoring exact orientation)
            path = []
            s = curr
            while s is not None:
                path.append(s)
                s = parent[s]
            path.reverse()
            return path, g[curr]

        for nbr, cost in successors(curr):
            if nbr in closed:
                continue
            new_g = g[curr] + cost
            if new_g < g.get(nbr, float("inf")):
                g[nbr] = new_g
                parent[nbr] = curr
                f_nbr = new_g + heuristic(nbr, goal_state)
                heapq.heappush(open_heap, (f_nbr, nbr))

    return None, float("inf")

# Example usage
start = (2, 2, angle_index(0.0))
goal = (40, 20, angle_index(0.0))
path, cost = astar(start, goal)
print("Path length:", len(path), "Total cost:", cost)
      
